### ISSUE TITLE
Provide XSL Transform of XML Results to HTML

### ISSUE THREAD

**Issue by rprouse** [nunit/nunit#294]
NUnit 2 ships with a simple XML -> Text XSL transform, but the results are nearly unreadable. They are just a wall of black text. Here is a small clipped example. This goes on for a thousand lines.

![nunit-transform](https://cloud.githubusercontent.com/assets/493828/4719770/4eb500ae-592c-11e4-9725-67b64d76c20f.png)

I am in the process of converting our unit tests over from MbUnit to NUnit 2, but the other devs are really pushing back, mainly because of the reports that are produced. MbUnit ships with really good XSL transforms which we use to produce the web reports from our continuous build server and they are included in the emails that get sent out with every build.

Developers also use them when running from the console. MbUnit has a **show report** command line switch and a **report type** switch (text, xml, html, xhtml, etc). The report is then generated in the desired format and launched using a process start to the default application.

Here are some examples of the reports that I've taken from our CI server,

![mbunit-header](https://cloud.githubusercontent.com/assets/493828/4719901/3352c84a-592d-11e4-920c-ff66a597f91c.png)

The details for the run contain a great deal of detail, but are easily readable. MbUnit ships with two transforms, full and condensed. Full shows all unit tests, condensed just shows ignored or failed tests. This example is of an ignored test, thus the grey bars as opposed to green.

Notice each section and level is collapsible, making it easier to work with the report.

![mbunit-details](https://cloud.githubusercontent.com/assets/493828/4719932/6f3bae58-592d-11e4-8d87-8c82611811ab.png)

Here is an example of a failed test. One cool feature is the ReSharper like error bar down the right side. The top shows the overall result, green or red, then below it are red lines for failed tests. You can click on those lines and they will take you to the failed tests. There are also grey lines for ignored tests.

Also notice that in the stack trace, user source files are linked. Gallio installs a handler for `gallio:` which will open the source file and navigate to the given line in Visual Studio. Of course this only works if your source is in the same location, but it is very handy when running from the GUI runner. The GUI runner uses an HTML control to display the report as the tests are run and this allows you to quickly jump to the source in Visual Studio for the failing tests.

`gallio:navigateTo?path=e:\Source\DotNet\TestingInfrastructure\TestFramework\FileAssert.cs&line=57`

![mbunit-failures](https://cloud.githubusercontent.com/assets/493828/4719991/dac928c6-592d-11e4-8cdb-778060f37672.png)

**Comment by CharliePoole** [comment]
@rprouse Are we ready to spell this out as a feature? It's listed in the 3.0 milestone, so we should either do that or move it to 'Future'. I have no preference.

**Comment by rprouse** [comment]
I would like to leave this one in for now, but I don't have a strong preference, so am willing to drop it. I will spell out here what I would like to see and we can discuss,

- NUnit ships with a simple XML to HTML transform. I think that once we provide a basic one, people will start extending it.
- Add a command line option to perform the transform and optionally specify your own transform and output filename
- Possibly add a Show Report command line option that does a process launch on the report and let's the OS launch the program for that extension type. If no transform is done, will show the XML report, otherwise the result of the transform. Will this work on Linux?

If we keep the HTML transform simple, all of this should be fairly easy to do and can be done in stages,

1. Create an HTML mockup of what a basic report should look like,
2. Create the XSLT to transform the XML to the HTML mockup.
3. Add the command line options to do the transform
4. Add the command line options to show the report

We could release at any point from 2 on and still have a viable feature.

**Comment by CharliePoole** [comment]
@yetibrain I see that I missed your post some months ago. Sorry I missed it. Have you tried using the results from our mock-assembly.exe?

**Comment by CharliePoole** [comment]
@rprouse Some comments on your outline...

 * I think the html transform should be a new format defined for the --result option. That is...
         --result=filename;format=html
   This will require a new addin to the engine.

 * The option to create the report using xslt already exists..
         --result=filename;format=user;transform=xsltfile

 * The command-line currently allows you to specify any number of output files in various formats, so the ShowReport option would need to be in some form that specified, which report to show. It may be simpler to just let folks launch the report they want manually. Alternatively, ShowReports could open the work directory in the file explorer, allowing the user to select the desired report. I see no particular problem doing this on Linux as well.

 * I think it's worth considering whether this should be part of nunit-console or a separate program. For example, I currently maintain (well, notionally anyway) a simple html report generator at https://github.com/CharliePoole/nunit-results I've been thinking of updating that and adding it to NUnit. I also have another one that uses XSLT, to do reports, although that's not my favorite way to do it. If we did this as a separate program, then it would be independent of the engine and could be updated by anyone. The console runner could have an option that launched the program, using the saved nunit3 format file.

 * On implementation, I think we need to decide whether to focus on creating the basic html report using xslt or C#. Personally, I find it trivial to do in C# and troublesome in XSLT. However, like you, I have done it both ways when I had to. Are we sure that users actually like XSLT better? Or are we leaning toward XSLT in order to fit in with what various CI servers use?
